pkg/hac/hac.go

```diff
@@ -0,0 +1,85 @@
+package hac
+
+// NooBaa Component Rescheduling
+// see doc/high-availability-controller.md
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/noobaa/noobaa-operator/v5/pkg/options"
+	"github.com/noobaa/noobaa-operator/v5/pkg/util"
+)
+
+// Name of this controller
+const Name = "high-availability-controller"
+
+// HAC is a high availability controller that watches the nodes Ready -> NotReady transitions
+// and deletes noobaa pods on the node that became unvailable
+type HAC struct {
+	Request  types.NamespacedName
+	Client   client.Client
+	Ctx      context.Context
+	Logger   *logrus.Entry
+	Node     *corev1.Node
+}
+
+// NewHAC initializes a high availability controller reconciler
+func NewHAC(
+	req types.NamespacedName,
+	client client.Client,
+) *HAC {
+	hac := &HAC{
+		Request:          req,
+		Client:           client,
+		Ctx:              context.Background(),
+		Logger:           logrus.WithField(Name, req.Name),
+	}
+
+	// the node that failed
+	hac.Node = &corev1.Node{}
+	hac.Node.Name = req.Name
+
+	return hac
+}
+
+// Reconcile is called when a node in the cluster transitions
+// from Ready to NotReady state
+// go over the noobaa pods running on this node and force their deletion
+func (hac *HAC) Reconcile() (reconcile.Result, error) {
+	res := reconcile.Result{}
+	log := hac.Logger
+	log.Warningf("❌ node %v became NotReady", hac.Node.Name)
+
+	// looking for noobaa pods running on the failing node in the watched namespace
+	labelOption := client.MatchingLabels{"app": "noobaa"}
+	namespaceOption := client.InNamespace(options.Namespace)
+	nodeOption := client.MatchingFields{"spec.nodeName": hac.Node.Name}
+
+	// fetch the noobaa pods from the api server
+	podList := &corev1.PodList{}
+	if !util.KubeList(podList, labelOption, namespaceOption, nodeOption) {
+		return res, errors.Errorf("failed to list noobaa pods on the node %v in namespace %v", hac.Node.Name, options.Namespace)
```

Comment:
when failing don't you need to requeue ?

Reply:
@romayalon 

> The Controller will requeue the Request to be processed again if an error is non-nil

From [here](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile#Reconciler), so returning non-nil error is enough.

Reply:
right, I mean that in other reconcilers we are setting the res.RequeueAfter to 3 seconds, do you think 3 seconds is an appropriate period for this reconciler too?

Reply:
@romayalon  not sure RequeueAfter should be specified. In other reconcilers, some processes indeed happen, say deployments are scaling or volumes are provisioned, etc... In this controller, the intent is to address the specific event of Node state transition. So unless there is an error no re-queue is ever needed.   if there is an error (pods list or pod delete) no requirements on specific delay from this controller. WDYT?

Reply:
no specific delay means default delay, is the default delay is appropriate for this reconciler?

Reply:
why would not it be? 😊

Reply:
The default period is ASAP, honoring RateLimit, see [here](https://github.com/kubernetes-sigs/controller-runtime/blob/985e819db7af1cd17782ca7ad27644d0ca655054/pkg/internal/controller/controller.go#L300-L304). LGTM.

Reply:
@baum AFAIK the default requeue is done with exponential backoff. maybe to better control the reconcile and make it more consistent we should use a constant delay

Reply:
@dannyzaken, by design, it is desired to handle the failure event immediately. Not sure how introducing artificial delay is beneficial here.